crates/ty_python_semantic/src/types/infer.rs

```diff
@@ -1676,10 +1679,25 @@ impl<'db, 'ast> TypeInferenceBuilder<'db, 'ast> {
                         // module, but that information is currently not threaded through attribute
                         // lookup.
                         if !previous_definition.kind(db).is_import() {
-                            let range = previous_definition.full_range(self.db(), self.module());
-                            diagnostic.annotate(
-                                self.context.secondary(range).message("Original definition"),
-                            );
+                            if let DefinitionKind::AnnotatedAssignment(assignment) =
+                                previous_definition.kind(db)
+                            {
+                                let range = assignment.annotation(self.module()).range();
+                                diagnostic.annotate(
+                                    self.context
+                                        .secondary(range)
+                                        .message("Symbol declared as `Final` here"),
+                                );
+                            } else {
+                                let range =
+                                    previous_definition.full_range(self.db(), self.module());
+                                diagnostic.annotate(
+                                    self.context
+                                        .secondary(range)
+                                        .message("Symbol declared as `Final` here"),
+                                );
```

Comment:
This is a fallback that currently can't be triggered, I believe, because there are no other annotated definitions that would allow a `Final` annotation. Function parameters are not allowed to be annotated with `Final`:

> Final may only be used in assignments or variable annotations. Using it in any other position is an error. In particular, Final can’t be used in annotations for function arguments

https://typing.python.org/en/latest/spec/qualifiers.html#uppercase-final

This is something that we still have to implement (added a note to the ticket for `typing.Final`).